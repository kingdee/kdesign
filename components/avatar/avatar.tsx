import React, { FunctionComponentElement, useContext } from 'react'
import classNames from 'classnames'
import ConfigContext from '../config-provider/ConfigContext'
import { tuple } from '../_utils/type'
import { getCompProps } from '../_utils'
import devWarning from '../_utils/devwarning'
import { Icon } from '../index'

export const AvatarSizes = tuple('large', 'middle', 'small', 'extra-small')
export type AvatarSize = typeof AvatarSizes[number]

export const AvatarShapes = tuple('circle', 'square')
export type AvatarShape = typeof AvatarShapes[number]

export interface AvatarProps {
  shape?: AvatarShape
  size?: AvatarSize | number
  gap?: number
  src?: React.ReactNode
  srcSet?: string
  draggable?: boolean
  icon?: React.ReactNode
  style?: React.CSSProperties
  prefixCls?: string
  className?: string
  children?: React.ReactNode
  alt?: string
  disabled?: boolean
  onError?: () => boolean
}

const InternalAvatar = (props: AvatarProps, ref: unknown): FunctionComponentElement<any> => {
  const { getPrefixCls, prefixCls, compDefaultProps: userDefaultProps } = useContext(ConfigContext)
  const avatarProps = getCompProps('Avatar', userDefaultProps, props)

  const {
    className,
    style,
    prefixCls: customPrefixcls,
    shape,
    size,
    icon,
    src,
    draggable,
    alt,
    children,
    gap,
    disabled,
    srcSet,
  } = avatarProps

  devWarning(
    AvatarSizes.indexOf(size) === -1 && typeof size !== 'number',
    'avatar',
    `cannot found avatar type '${size}'`,
  )

  devWarning(AvatarShapes.indexOf(shape) === -1, 'avatar', `cannot found avatar shape '${shape}'`)

  const [isImgExist, setIsImgExist] = React.useState(true)
  const [scale, setScale] = React.useState(1)
  const [mounted, setMounted] = React.useState(false)

  // ref
  const avatarRef = (ref as any) || React.createRef<HTMLElement>()
  const avatarChildrenRef = React.useRef<HTMLElement>()

  // class
  const avatarPrefixCls = getPrefixCls!(prefixCls, 'avatar', customPrefixcls)
  const sizeClasses = classNames({
    [`${avatarPrefixCls}-lg`]: size === 'large',
    [`${avatarPrefixCls}-md`]: size === 'middle',
    [`${avatarPrefixCls}-sm`]: size === 'small',
    [`${avatarPrefixCls}-xs`]: size === 'extra-small',
  })

  const hasImageElement = React.isValidElement(src)

  const isShowDefaultIcon = !src && !icon && !children

  const avatarClasses = classNames(
    avatarPrefixCls,
    sizeClasses,
    {
      [`${avatarPrefixCls}-${shape}`]: !!shape,
      [`${avatarPrefixCls}-image`]: hasImageElement || (src && isImgExist),
      [`${avatarPrefixCls}-icon`]: !!icon || isShowDefaultIcon,
      [`${avatarPrefixCls}-disabled`]: disabled,
    },
    className,
  )

  const sizeStyle: React.CSSProperties =
    typeof size === 'number'
      ? {
          width: size,
          height: size,
          lineHeight: `${size}px`,
          fontSize: icon ? size / 2 : 18,
        }
      : {}

  React.useEffect(() => {
    setIsImgExist(true)
    setScale(1)
  }, [src])

  React.useEffect(() => {
    setScaleParam()
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [gap])

  React.useEffect(() => {
    setMounted(true)
  }, [])

  const handleImgLoadError = () => {
    const { onError } = props
    const errorFlag = onError ? onError() : undefined
    if (errorFlag !== false) {
      setIsImgExist(false)
    }
  }

  const setScaleParam = () => {
    if (!avatarChildrenRef.current || !avatarRef.current) {
      return
    }
    const childrenWidth = avatarChildrenRef.current.offsetWidth
    const nodeWidth = avatarRef.current.offsetWidth
    if (childrenWidth !== 0 && nodeWidth !== 0) {
      if (gap * 2 < nodeWidth) {
        setScale(nodeWidth - gap * 2 < childrenWidth ? (nodeWidth - gap * 2) / childrenWidth : 1)
      }
    }
  }

  let childrenToRender
  if (typeof src === 'string' && isImgExist) {
    childrenToRender = <img src={src} srcSet={srcSet} draggable={draggable} onError={handleImgLoadError} alt={alt} />
  } else if (hasImageElement) {
    childrenToRender = src
  } else if (icon) {
    childrenToRender = icon
  } else if (!children) {
    childrenToRender = <Icon type="avatar-solid" />
  } else if (mounted || scale !== 1) {
    const childrenStyle: React.CSSProperties = {
      transform: `scale(${scale}) translateX(-50%)`,
    }
    const sizeChildrenStyle: React.CSSProperties =
      typeof size === 'number'
        ? {
            lineHeight: `${size}px`,
          }
        : {}

    childrenToRender = (
      <span
        className={`${avatarPrefixCls}-string`}
        ref={(node: HTMLElement) => {
          avatarChildrenRef.current = node
        }}
        style={{ ...sizeChildrenStyle, ...childrenStyle }}
      >
        {children}
      </span>
    )
  } else {
    childrenToRender = (
      <span
        className={`${avatarPrefixCls}-string`}
        style={{ opacity: 0 }}
        ref={(node: HTMLElement) => {
          avatarChildrenRef.current = node
        }}
      >
        {children}
      </span>
    )
  }

  return (
    <span ref={avatarRef} className={avatarClasses} style={{ ...sizeStyle, ...style }}>
      {childrenToRender}
    </span>
  )
}

const Avatar = React.forwardRef<unknown, AvatarProps>(InternalAvatar)
Avatar.displayName = 'Avatar'
export default Avatar
