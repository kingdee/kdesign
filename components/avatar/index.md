---
category: Components
type: 展示
subtitle: 头像
title: Avatar
order: 0
---

头像是一种代表人或物的组件。

## 使用场景

需要代表某个人或物时使用，例如：用户、商家、店铺和企业等。

## API

| 属性 | 说明 | 类型 | 默认值 | 版本 |
| --- | --- | --- | --- | --- |
| alt | 图像无法显示时的替代文本 | string | - | 1.0.0 |
| draggable | 图片是否允许拖动 | boolean \| `'true'` \| `'false'` | - | 1.0.0 |
| gap | 字符类型距离左右两侧边界单位像素 | number | 4 | 1.0.0 |
| icon | 设置头像的自定义图标 | ReactNode | - | 1.0.0 |
| onError | 图片加载失败的事件，返回 false 会关闭组件默认的 fallback 行为 | () => boolean | - | 1.0.0 |
| shape | 指定头像的形状 | `circle` \| `square` | `circle` | 1.0.0 |
| size | 设置头像的大小 | number \| `large` \| `middle` \| `small` \| `extra-small` | `middle` | 1.0.0 |
| src | 图片类头像的资源地址或者图片元素 | string \| ReactNode | - | 1.0.0 |
| srcSet | 设置图片类头像响应式资源地址 | string | - | 1.0.0 |

## Design Token

| 分类 | 组件token | 全局token | 默认值 |
| --- | --- | --- | --- |
| color | --kd-c-avatar-color-background | - | #B2B2B2 |
| font | --kd-c-avatar-font-size-x-small | - | 14px |
|  | --kd-c-avatar-font-size-small | - | 18px |
|  | --kd-c-avatar-font-size-base | - | 22px |
|  | --kd-c-avatar-font-size-large | - | 36px |
| radius | --kd-c-avatar-border-radius | - | 2px |
| sizing | --kd-c-avatar-size-x-small | - | 24px |
|  | --kd-c-avatar-size-small | - | 32px |
|  | --kd-c-avatar-size-base | - | 40px |
|  | --kd-c-avatar-size-large | - | 60px |
