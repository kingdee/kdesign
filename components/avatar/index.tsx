import Avatar from './avatar'
export * from './avatar'
export default Avatar
export type { AvatarProps as IAvatarProps } from './avatar'
