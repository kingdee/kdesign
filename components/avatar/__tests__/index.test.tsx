import React from 'react'
import { mount, render } from 'enzyme'
import ConfigProvider from '../../config-provider/index'
import Avatar from '../index'
import { AvatarShapes, AvatarShape, AvatarSizes } from '../avatar'
import Icon from '../../icon'
import mountTest from '../../../tests/shared/mountTest'

describe('Avatar', () => {
  // 1.mount test
  describe('1.mount test', () => {
    mountTest(Avatar)
    AvatarShapes.forEach((shape) => {
      mountTest(() => <Avatar shape={shape} />)
    })
    AvatarSizes.forEach((size) => {
      mountTest(() => <Avatar size={size} />)
    })
  })

  // 2.render test
  describe('2.render test', () => {
    it('renders correctly', () => {
      expect(render(<Avatar>Avatar</Avatar>)).toMatchSnapshot()
      AvatarSizes.forEach((size) => {
        expect(render(<Avatar size={size}>Avatar</Avatar>)).toMatchSnapshot()
      })
      AvatarShapes.forEach((shape) => {
        expect(render(<Avatar shape={shape}>Avatar</Avatar>)).toMatchSnapshot()
      })
    })
  })

  // 3.warns in component
  describe('3.warns in component', () => {
    it('warns if type is wrong', () => {
      const mockWarn = jest.fn()
      jest.spyOn(console, 'warn').mockImplementation(mockWarn)
      const shape = 'oval' as any as AvatarShape
      render(<Avatar shape={shape} />)
      expect(mockWarn).toHaveBeenCalledTimes(1)
      expect(mockWarn.mock.calls[0][0]).toMatch("Warning: [kdesign]-avatar: cannot found avatar shape 'oval'")
    })
  })
  // 4. render null or undefined without errors
  describe('4. render null or undefined without errors', () => {
    it('render null or undefined buttons without errors', () => {
      expect(
        mount(
          <Avatar>
            {null}
            {undefined}
          </Avatar>,
        ),
      ).toMatchSnapshot()
    })
  })

  // 5. displayName
  describe('5. displayName', () => {
    it('should have displayName static property', () => {
      const wrapper = mount(<Avatar>Avatar</Avatar>)
      expect((wrapper.type() as any).displayName).toBe('Avatar')
    })
  })
  // 6. class state
  describe('6. class state', () => {
    it('should class use right', () => {
      // default
      const DefaultAvatar = mount(<Avatar />)
      expect(DefaultAvatar.find('.kd-avatar')).toHaveClassName('.kd-avatar-md')
      expect(DefaultAvatar.find('.kd-avatar')).toHaveClassName('.kd-avatar-circle')

      const stringAvatar = mount(<Avatar>string</Avatar>)
      expect(stringAvatar.find('.kd-avatar').length).toBeGreaterThanOrEqual(1)

      AvatarShapes.forEach((shape) => {
        const TestButton = mount(<Avatar shape={shape}>Avatar</Avatar>)
        expect(TestButton.find(`.kd-avatar`)).toHaveClassName(`.kd-avatar-${shape}`)
      })

      const DisableAvatar = mount(<Avatar disabled />)
      expect(DisableAvatar.find(`.kd-avatar`)).toHaveClassName('.kd-avatar-disabled')

      const IconAvatar = mount(<Avatar icon={<Icon type="add"></Icon>} />)
      expect(IconAvatar.find(`.kd-avatar`)).toContainReact(<Icon type="add"></Icon>)
    })
  })

  // 7.component interaction(event)
  describe('7.component interaction(event)', () => {
    it('Render long string correctly', () => {
      const wrapper = mount(<Avatar>TestString</Avatar>)
      const children = wrapper.find('.kd-avatar-string')
      expect(children.length).toBe(1)
    })

    it('should render fallback string correctly', () => {
      const div = global.document.createElement('div')
      global.document.body.appendChild(div)

      const wrapper = mount(<Avatar src="http://error.url">Fallback</Avatar>, { attachTo: div })
      wrapper.find('img').simulate('error')
      const children = wrapper.find('.kd-avatar-string')
      expect(children.length).toBe(1)
      expect(children.text()).toBe('Fallback')
      wrapper.detach()
      global.document.body.removeChild(div)
    })

    it('should handle onError correctly', () => {
      const LOAD_FAILURE_SRC = 'http://error.url'
      const LOAD_SUCCESS_SRC = 'https://kui.kingdee.com/assets/image/avatar_m.png'

      const div = global.document.createElement('div')
      global.document.body.appendChild(div)

      class Foo extends React.Component {
        state = {
          src: LOAD_FAILURE_SRC,
        }

        handleImgError = () => {
          this.setState({
            src: LOAD_SUCCESS_SRC,
          })
          return false
        }

        render() {
          const { src } = this.state
          return <Avatar src={src} onError={this.handleImgError} />
        }
      }

      const wrapper = mount(<Foo />, { attachTo: div })
      expect(div.querySelector('img')?.getAttribute('src')).toBe(LOAD_FAILURE_SRC)
      wrapper.find('img').simulate('error')

      expect(wrapper).toMatchSnapshot()
      expect(div.querySelector('img')?.getAttribute('src')).toBe(LOAD_SUCCESS_SRC)

      wrapper.detach()
      global.document.body.removeChild(div)
    })
  })

  // 8.config provider
  describe('8.config provider', () => {
    it('should config use config provider', () => {
      const buttonConfig = {
        compDefaultProps: {
          Avatar: {
            size: 'large',
          },
        },
      }
      const wrapper = mount(
        <ConfigProvider value={buttonConfig}>
          <Avatar />
        </ConfigProvider>,
      )
      expect(wrapper.find('.kd-avatar')).toHaveClassName('.kd-avatar-lg')
    })
  })

  // 9. ref test
  describe('9. ref test', () => {
    it('should get avatar element from ref', () => {
      const ref = React.createRef()
      mount(<Avatar ref={ref}></Avatar>)
      expect(ref.current instanceof HTMLSpanElement).toBe(true)
    })
  })

  // 10.api test
  describe('10.api test', () => {
    it('alt', () => {
      // alt
      const altAvatar = mount(<Avatar alt="替代" src="https://kui.kingdee.com/assets/image/avatar_m.png" />)
      expect(altAvatar.find(`.kd-avatar`).find('img').prop('alt')).toEqual('替代')
    })
    it('draggable', () => {
      // draggable
      const draggableAvatar = mount(
        <Avatar draggable={false} src="https://kui.kingdee.com/assets/image/avatar_m.png" />,
      )
      expect(draggableAvatar.find(`.kd-avatar`).find('img').prop('draggable')).toEqual(false)
    })
    it('icon', () => {
      // icon
      const iconAvatar = mount(<Avatar icon={<Icon type="loadding" spin />} style={{ color: '#ffffff' }} />)
      expect(iconAvatar.find(`.kdicon-loadding`).length).toBe(1)
    })
    it('srcSet', () => {
      const srcSetAvatar = mount(
        <Avatar
          src="https://kui.kingdee.com/assets/image/avatar_m.png"
          srcSet="https://kui.kingdee.com/assets/image/avatar_m.png"
          onError={() => false}
        />,
      )
      expect(srcSetAvatar.find(`.kd-avatar`).find('img').prop('srcSet')).toEqual(
        'https://kui.kingdee.com/assets/image/avatar_m.png',
      )
    })
  })
})
