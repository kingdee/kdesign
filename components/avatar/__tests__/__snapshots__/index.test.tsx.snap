// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar 2.render test renders correctly 1`] = `
<span
  class="kd-avatar kd-avatar-md kd-avatar-circle"
>
  <span
    class="kd-avatar-string"
    style="opacity:0"
  >
    Avatar
  </span>
</span>
`;

exports[`Avatar 2.render test renders correctly 2`] = `
<span
  class="kd-avatar kd-avatar-lg kd-avatar-circle"
>
  <span
    class="kd-avatar-string"
    style="opacity:0"
  >
    Avatar
  </span>
</span>
`;

exports[`Avatar 2.render test renders correctly 3`] = `
<span
  class="kd-avatar kd-avatar-md kd-avatar-circle"
>
  <span
    class="kd-avatar-string"
    style="opacity:0"
  >
    Avatar
  </span>
</span>
`;

exports[`Avatar 2.render test renders correctly 4`] = `
<span
  class="kd-avatar kd-avatar-sm kd-avatar-circle"
>
  <span
    class="kd-avatar-string"
    style="opacity:0"
  >
    Avatar
  </span>
</span>
`;

exports[`Avatar 2.render test renders correctly 5`] = `
<span
  class="kd-avatar kd-avatar-xs kd-avatar-circle"
>
  <span
    class="kd-avatar-string"
    style="opacity:0"
  >
    Avatar
  </span>
</span>
`;

exports[`Avatar 2.render test renders correctly 6`] = `
<span
  class="kd-avatar kd-avatar-md kd-avatar-circle"
>
  <span
    class="kd-avatar-string"
    style="opacity:0"
  >
    Avatar
  </span>
</span>
`;

exports[`Avatar 2.render test renders correctly 7`] = `
<span
  class="kd-avatar kd-avatar-md kd-avatar-square"
>
  <span
    class="kd-avatar-string"
    style="opacity:0"
  >
    Avatar
  </span>
</span>
`;

exports[`Avatar 4. render null or undefined without errors render null or undefined buttons without errors 1`] = `
<Avatar>
  <span
    className="kd-avatar kd-avatar-md kd-avatar-circle"
    style={Object {}}
  >
    <span
      className="kd-avatar-string"
      style={
        Object {
          "transform": "scale(1) translateX(-50%)",
        }
      }
    />
  </span>
</Avatar>
`;

exports[`Avatar 7.component interaction(event) should handle onError correctly 1`] = `
<Foo>
  <Avatar
    onError={[Function]}
    src="https://kui.kingdee.com/assets/image/avatar_m.png"
  >
    <span
      className="kd-avatar kd-avatar-md kd-avatar-circle kd-avatar-image"
      style={Object {}}
    >
      <img
        onError={[Function]}
        src="https://kui.kingdee.com/assets/image/avatar_m.png"
      />
    </span>
  </Avatar>
</Foo>
`;
